deltacat/utils/ray_utils/retry_handler/retry.py

```diff
@@ -0,0 +1,175 @@
+
+
+
+
+class RetryHandler:
+    #given a list of tasks that are failing, we want to classify the error messages and redirect the task
+    #depending on the exception type using a wrapper
+    #wrapper function that before execution, checks what exception is being thrown and go to second method to
+    #commence retrying
+    def execute_task(self, ray_remote_task_info: RayRemoteTaskInfo) -> Any:
+            self.start_tasks_execution([ray_remote_task_info])
+            return self.wait_and_get_all_task_results()[0]
+
+    """
+    Starts execution of all given Ray remote tasks
+    """
+    def start_tasks_execution(self, ray_remote_task_infos: List[TaskInfoObject]) -> None:
+        self.start_tasks_execution_in_batches(ray_remote_task_infos, RayRemoteTasksBatchScalingParams(initial_batch_size=len(ray_remote_task_infos)))
+
+    """
+    Starts execution of given Ray remote tasks in batches depending on the given Batch scaling params
+    """
+    def start_tasks_execution_in_batches(self, ray_remote_task_infos: List[RayRemoteTaskInfo], batch_scaling_params: RayRemoteTasksBatchScalingParams) -> None:
+        self.num_of_submitted_tasks = len(ray_remote_task_infos)
+        self.current_batch_size = min(batch_scaling_params.initial_batch_size, self.num_of_submitted_tasks)
+        self.num_of_submitted_tasks_completed = 0
+        self.remaining_ray_remote_task_infos = ray_remote_task_infos
+        self.batch_scaling_params = batch_scaling_params
+        self.task_promise_obj_ref_to_task_info_map: Dict[Any, RayRemoteTaskInfo] = {}
+
+        self.unfinished_promises: List[Any] = []
+        logger.info(f"Starting the execution of {len(ray_remote_task_infos)} Ray remote tasks. Concurrency of tasks execution: {self.current_batch_size}")
+        self.__submit_tasks(self.remaining_ray_remote_task_infos[:self.current_batch_size])
+        self.remaining_ray_remote_task_infos = self.remaining_ray_remote_task_infos[self.current_batch_size:]
+
+
+    def wait_and_get_all_task_results(self) -> List[Any]:
+        return self.wait_and_get_task_results(self.num_of_submitted_tasks)
+
+    def get_task_results(self, num_of_results: int) -> List[Any]:
+    #implement wrapper here that before execution will try catch an exception
+        #get what tasks we need to run our execution on
+        finished, unfinished = ray.wait(unfinished, num_of_results)
+        #assuming we have the tasks we want to get results of
+        for finished in finished:
+            finished_result = None
+            try:
+                finished_result = ray.get(finished)
+            except (Exception) as exception:
+                #if exception send to method handle_ray_exception to determine what to do and assign the corresp error
+                finished_result = self.handle_ray_exception(exception=exception, TaskInfoObject = )#evaluate the exception and return the error
+
+            if finished_result and type(finished_result) == RayRemoteTaskExecutionError:
```

Comment:
The interface is still not clear. Can you add a README on how to use this class to accomplish a) notifying progress, b) detecting stragglers, c) retrying on exceptions that are retryable.